Fixed some typos and errors in the documentation and added new doc page. (#88)
front-commerce/developers.front-commerce.com: source/docs/magento2/add-new-attribute.md

---
id: add-new-attribute
title: Add a new attribute
---

Let's say we want to add a new element to display on a product's page (for instance, a product's rate). You will have to:

1. Add the attribute's actual value to your database.
2. Create a new module.
3. Define the schema, implement the resolver and display it on screen.

## Add a value to your database

> The process to do so may vary with the backend software you are using. For Magento 2 (used in this example), you can refer to [this](https://www.fastcomet.com/tutorials/magento2/product-attributes).

Reviewer: ```suggestion
The process to do so may vary with the backend software you are using. For Magento 2 (used in this example), you can refer to [How to Create Product Attributes in Magento 2(https://www.fastcomet.com/tutorials/magento2/product-attributes).
```

Links saying stuff like `click here` are not accessible to everyone. Stating exactly what the link is about within the link will improve accessibility and make things more explicit.

## Create a new GraphQL module

> We want to display the `rate` of a product on the product page. To keep it simple, it will only consist in displaying a number, the value can be manually edited by the merchant. This feature will be stored in a new module called `mymodule` ; you could directly edit the files at `node_modules/front-commerce/path/to/something`, but this is not recommended since running `npm install` might erase any modification.

Reviewer: ```suggestion
We want to display the `rate` of a product on the product page. To keep it simple, it will only consist in displaying a number, which can be manually edited by the merchant. This feature will be stored in a new module called `mymodule`.

<blockquote class="info">Instead of creating a module, you could directly edit the files at `node_modules/front-commerce/path/to/something`. However it is highly discouraged since running `npm install` might erase any modification.</blockquote>
```

I've moved the note about the node_modules in a separate paragraph because it is a best practice in the `node` ecosystem, so it's not mandatory to mention it. Thus I make this a tiny bit more explicit by changing the information hierarchy.

> Learn more about Front-Commerce's folder structure [here](/docs/essentials/extend-the-theme.html).

Reviewer: In my previous comment I might have done a mistake. In this part, it's `/docs/essentials/extend-the-graphql-schema.html` which is important.

And same thing about a link having `here` as text. How would you replace it?

```bash
mkdir mymodule
cd mymodule
```

In this example, the files we are going to override are located at `node_modules/front-commerce/src/web/theme/modules/ProductView/Synthesis/`, so will create in our `mymodule` folder the following directories. More details [here](/docs/essentials/extend-the-graphql-schema.html).

```bash

> mkdir -p server/modules/acme

Reviewer: ```suggestion
mkdir -p mymodule/server/modules/ratings
```

Let's make explicit that this should live in `mymodule`. Plus, `acme` is actually not a good name for our extension since it's only made for adding `ratings` functionality to the GraphQL server. Naming it will help people navigate in their codebase.

And this is an opportunity to change the fact that `acme` was `chosen arbitrarily`.

```

(Note that the `acme` name has been chosen arbitrarily.)

## Add the module to Front-Commerce

Now we will add the module we just made to Front-Commerce. Edit the `.front-commerce.js` file (root folder) as shown below to do so.

```diff
module.exports = {
  name: "Front Commerce Skeleton",
  url: "http://localhost:4000",
- modules: ["./src"],
+ modules: ["./mymodule", "./src"],
  serverModules: [
    { name: "FrontCommerce", path: "server/modules/front-commerce" },
-   { name: "Magento2", path: "server/modules/magento2" }
+   { name: "Magento2", path: "server/modules/magento2" },
+   { name: "Acme", path: "./mymodule/server/modules/acme" }
  ]
};
```

### Expand the Product definition

> Create a file named `schema.gql` and type in the following code. It means that we are expanding the definition of a product: it can now be rated.

Reviewer: I'd add an introduction like:

> First, we need to tell the server how to customize the GraphQL schema.

Otherwise we don't really know why you're creating a `schema.gql`, in my opinion.

```js
// mymodule/server/modules/acme/schema.gql
extend type Product {
  rate: Float
}
```

### Implement the resolver

We must now define what Front-Commerce should fetch when `rate` is requested in a GraphQL query; in other words we must write the code that will **resolve** the queries for the rates.

Create a file named `resolvers.js` and type in the following code. It implements the resolver.

```js
// mymodule/server/modules/acme/resolvers.js
export default {
  Product: {
    rate: parent => {

>       return parseFloat(parent.rate);

Reviewer: Why did you use the `parent` directly?

Maybe you could add a paragraph explaining that by default all the attributes are available in a product. It's just that they are not exposed on the GraphQL schema yet.

Author: Is it better like this ? I hadn't thought thought of destructuring before.

```js
// mymodule/server/modules/ratings/resolvers.js
export default {
  Product: {
    rate: {rate} => parseFloat(rate);
  }
};
```

By default all the attributes are available in a product. It's just that they are not exposed on the GraphQL schema yet.

    }
  }
};
```

### Declare the module

If we want our code to be taken into account when the module is loaded, we must reference it.

In the file named `index.js`, type in the following code. It references the schema and the resolver from earlier.

```js
// mymodule/server/modules/acme/index.js
import typeDefs from "./schema.gql";
import resolvers from "./resolvers";

export default {
  namespace: "Acme",
  typeDefs: typeDefs,
  resolvers: resolvers
};
```

## Discover the playground

> By typing `yourhostname/playground` or `yourhostname/graphiql` (in our case `localhost:4000/playground`) in your browser address bar, you can access a GraphQL playground with a nice GUI to test your queries. For instance, type the following code in the left pane and press `Ctrl` + `Enter` (or click the arrow button)

Reviewer: Don't forget to mention the fact that one need to restart their application before accessing to the updated schema in the playground.

> By typing `yourhostname/playground` or `yourhostname/graphiql` (in our case `localhost:4000/playground`) in your browser address bar, you can access a GraphQL playground with a nice GUI to test your queries. For instance, type the following code in the left pane and press `Ctrl` + `Enter` (or click the arrow button)

Reviewer: Is it an `arrow button` or a `play button`?

```graphql
{
  product(sku: "yourSKUhere") {
    name
    rate
  }
}
```

From this GraphQL query, you should get a JSON content that looks like this:

```json
{
  "data": {
    "product": {
      "name": "Your product name",
      "rate": // the value stored in your backoffice (Magento 2 in this case)
    }
  }
}
```

### Overriding the files

> In **this** example, the files to be overridden are `ProductSynthesisFragment.gql` and `Synthesis.js` both located at `node_modules/front-commerce/src/web/theme/modules/ProductView/Synthesis/`. Copy and paste them into `mymodule/web/theme/modules/ProductView/Synthesis`.

Reviewer: I think that this paragraph should be after `How to find which files to edit`.

Author: Is it better like this ? I hadn't thought thought of destructuring before.

```js
// mymodule/server/modules/ratings/resolvers.js
export default {
  Product: {
    rate: {rate} => {
      return parseFloat(rate);
    }
  }
};
```

By default all the attributes are available in a product. It's just that they are not exposed on the GraphQL schema yet.

#### How to find which files to edit

> On your web page, right click on the element you wish to modify, then click on **Inspect Element**. In the inspector, you can look for the `className` that holds the elements you want to edit. Then you can search for that keyword in your `node_modules/front-commerce/src` folder and you will find the right files.

Reviewer: ```suggestion
On your web page, right click on the element you wish to modify, then click on **Inspect Element**. In the inspector, you can look for the `class` that holds the elements you want to edit. Then you can search for that keyword in your `node_modules/front-commerce/src` folder and you will find the right files.
```

> We've included images to illustrate our example.

Reviewer: Rather than saying everything in one paragraph and then adding images, I'd suggest to split the paragraph and put the images exactly when you are talking about them: 1. part about the inspector, 2. part about the search in your files.

> ![Inspector screenshot here](./assets/inspector-screenshot.png)

Reviewer: The alt name is not really relevant here. The goal of the `alt` attribute is to explain to people that can't display or see the image what it's illustrating.

Here is a suggestion.

```suggestion
![Inspector screenshot demonstrating that when you inspect a product page description, you find the product__sythesis class in the DOM](./assets/inspector-screenshot.png)
```

> ![Synthesis screenshot here](./assets/synthesis-screenshot.png)

Reviewer: How would you change the alt name here?

### Updating the query

> In `ProductSynthesisFragment.gql`, add the following query line. It means that the application will request the `rate` field as well when sending the query. (Depending on the version of Front-Commerce you are using, the content might differ slightly.)

Reviewer: Why did you update the `ProductSynthesisFragment.gql`? It'd be nice to explain that GraphQL queries can be splitted in Fragments and that each fragment is supposed to live next to a component in order to easily request all the data needed for your component. Thus, since we need to display some new data on the ProductSynthesis component, we update the ProductSynthesisFragment.

```diff

> #import "theme/components/organisms/Configurator/ProductConfiguratorFragment.gql"

Reviewer: In most of your code examples, there was a `// path/to/file` at the begginning. We should do the same here.

Reviewer: This comment was not resolved here.

#import "theme/modules/AddToCart/ProductStockFragment.gql"
#import "theme/components/atoms/Typography/Price/ProductPriceFragment.gql"
#import "theme/modules/ProductView/Details/ProductDetailsFragment.gql"

fragment ProductSynthesisFragment on Product {
  sku
  name
  ...ProductPriceFragment
  description
+ rate
  ...ProductConfiguratorFragment
  ...ProductStockFragment
  ...ProductDetailsFragment
  configurations {
    product {
      sku
      ...ProductPriceFragment
    }
  }
}
```

### Displaying the result on screen

We're almost there! In the file `Synthesis.js` (which generates the product page), you can add some code to display the `rate`. Knowing React.js or at least HTML syntax is necessary to perform this step without impacting other items on the page.

```diff
// Many more things here
const ProductSynthesis = props => {
  const {
    product
    // Many more things here
  } = props;
  return (
    <div className="product__synthesis">
      {/* Many more things here */}
+     <div>Rate: {product.rate} / 5</div>
      {/* Many more things here */}
    </div>
  );
};

export default ProductSynthesis;

> ```

Reviewer: It would be great to have a final screenshot here to show the expected result.

> ```

Reviewer: Could you please add a conclusion about this documentation?

A lead could be to suggest to the user to add a custom attribute to a category. This will be the exact same steps as here and will help them to make sure they've understood the concepts correctly.

---
id: add-new-attribute
title: Add a new attribute
---

Let's say we want to add a new element to display on a product's page (for instance, a product's rate). You will have to:

1. Add an attribute to your Magento 2.
2. Expose the attribute in your GraphQL schema.
3. Display it on screen.

## Add a value to your database

The process to do so may vary with the backend software you are using. For Magento 2 (used in this example), you can refer to [How to Create Product Attributes in Magento 2](https://www.fastcomet.com/tutorials/magento2/product-attributes).

## Expose the attribute in your GraphQL schema

We want to display the `rate` of a product on the product page. To keep it simple, it will only consist in displaying a number, which can be manually edited by the merchant. This feature will be stored in a new module called `mymodule`.

<blockquote class="info">
You could directly edit the files at `node_modules/front-commerce/path/to/something`, but this is not recommended since running `npm install` might erase any modification.
Learn [how to extend the GraphQL schema](/docs/essentials/extend-the-graphql-schema.html).
</blockquote>

```bash
mkdir mymodule
```

> In this example, the files we are going to override are located at `node_modules/front-commerce/src/web/theme/modules/ProductView/Synthesis/`, so will create in our `mymodule` folder the following directories. More details [here](/docs/essentials/extend-the-graphql-schema.html).

Reviewer: In this sentence, you talk about a React component and yet you're talking about the server in the following command.

I think that this sentence should be updated. The React Component is only relevant in the `How to find which files to edit` part.

```bash
mkdir -p mymodule/server/modules/ratings
```

> ## Add the module to Front-Commerce

Reviewer: ```suggestion
### Add the module to Front-Commerce
```

```diff
module.exports = {
  name: "Front Commerce Skeleton",
  url: "http://localhost:4000",
- modules: ["./src"],
+ modules: ["./mymodule", "./src"],
  serverModules: [
    { name: "FrontCommerce", path: "server/modules/front-commerce" },
-   { name: "Magento2", path: "server/modules/magento2" }
+   { name: "Magento2", path: "server/modules/magento2" },
+   { name: "Ratings", path: "./mymodule/server/modules/ratings" }
  ]
};
```

### Expand the Product definition

First, we need to tell the server how to customize the GraphQL schema. To do so, we will create a file named `schema.gql` and type in the following code. It means that we are expanding the definition of a product: it can now be rated.

```js
// mymodule/server/modules/ratings/schema.gql
extend type Product {
  rate: Float
}
```

### Implement the resolver

We must now define what Front-Commerce should fetch when `rate` is requested in a GraphQL query; in other words we must write the code that will **resolve** the queries for the rates.

Create a file named `resolvers.js` and type in the following code. It implements the resolver.

```js
// mymodule/server/modules/ratings/resolvers.js
export default {
  Product: {

>     rate: {rate} => parseFloat(rate);

Reviewer: This code is not valid.

```suggestion
    rate: ({rate}) => parseFloat(rate)
```

Author: I noticed that too.

  }
};
```

By default all the attributes are available in a product. It's just that they are not exposed on the GraphQL schema yet.

### Declare the module

If we want our code to be taken into account when the module is loaded, we must reference it.

> In the file named `index.js`, type in the following code. It references the schema and the resolver from earlier.

Reviewer: ```suggestion
In the file named `index.js`, type the following code. It references the schema and the resolver from earlier.
```

```js
// mymodule/server/modules/ratings/index.js
import typeDefs from "./schema.gql";
import resolvers from "./resolvers";

export default {
  namespace: "Ratings",
  typeDefs: typeDefs,
  resolvers: resolvers
};
```

### Discover the playground

> By typing `yourhostname/playground` or `yourhostname/graphiql` (in our case `localhost:4000/playground`) in your browser address bar, you can access a GraphQL playground with a nice GUI to test your queries. For instance, type the following code in the left pane and press `Ctrl` + `Enter` (or click the **play** button). You will need to restart the application to access the updated schema in the playground.

Reviewer: ```suggestion
By typing `yourhostname/playground` or `yourhostname/graphiql` (in our case `localhost:4000/playground`) in your browser address bar, you can access a GraphQL playground with a nice GUI to test your queries. For instance, type the following code in the left pane and press `Ctrl` + `Enter` (or click the **play** button).

**You will need to restart the application to access the updated schema in the playground.**
```

```graphql
{
  product(sku: "yourSKUhere") {
    name
    rate
  }
}
```

From this GraphQL query, you should get a JSON content that looks like this:

```json
{
  "data": {
    "product": {
      "name": "Your product name",
      "rate": // the value stored in your backoffice (Magento 2 in this case)
    }
  }
}
```

> ### How to find which files to edit

Reviewer: ```suggestion
## Display the attribute in a product page

Now that you have access to your attribute, you can display it however you want in your React application. In this example, we will add it on the product's page, in the right section.

### How to find which files to edit
```

> On your web page, right click on the element you wish to modify, then click on **Inspect Element**. In the inspector, you can look for the `class` that holds the elements you want to edit. Here, the `class` is `product__synthesis`.

Reviewer: ```suggestion
On your web page, right click on the element you wish to edit, then click on **Inspect Element**. In the inspector, you can look for the `class` that holds the elements you want to edit. Here, the `class` is `product__synthesis`.
```

![Alt: Inspector screenshot demonstrating that when you inspect a product page description, you find the product__sythesis class in the DOM](./assets/inspector-screenshot.png)

Then you can search for that keyword in your `node_modules/front-commerce/src` folder and you will find the right files.

![Alt: Synthesis screenshot that shows where to find the files given a `class` value](./assets/synthesis-screenshot.png)

### Overriding the files

> In **this** example, the files to be overridden are `ProductSynthesisFragment.gql` and `Synthesis.js` both located at `node_modules/front-commerce/src/web/theme/modules/ProductView/Synthesis/`. Copy and paste them into `mymodule/web/theme/modules/ProductView/Synthesis`.

Reviewer: ```suggestion
In **this** example, the files to be overridden are `ProductSynthesisFragment.gql` and `Synthesis.js` both located at `node_modules/front-commerce/src/web/theme/modules/ProductView/Synthesis/`. Copy and paste them into `mymodule/web/theme/modules/ProductView/Synthesis`. See [Extend the theme](/docs/essentials/extend-the-theme.html) for more details.
```

### Updating the query

In `mymodule/web/theme/modules/ProductView/Synthesis/ProductSynthesisFragment.gql`, add the following query line. It means that the application will request the `rate` field as well when sending the query. (Depending on the version of Front-Commerce you are using, the content might differ slightly.)

```diff
#import "theme/components/organisms/Configurator/ProductConfiguratorFragment.gql"
#import "theme/modules/AddToCart/ProductStockFragment.gql"
#import "theme/components/atoms/Typography/Price/ProductPriceFragment.gql"
#import "theme/modules/ProductView/Details/ProductDetailsFragment.gql"

fragment ProductSynthesisFragment on Product {
  sku
  name
  ...ProductPriceFragment
  description
+ rate
  ...ProductConfiguratorFragment
  ...ProductStockFragment
  ...ProductDetailsFragment
  configurations {
    product {
      sku
      ...ProductPriceFragment
    }
  }
}
```

<blockquote class="info">

> GraphQL queries can be split in Fragments and that each fragment is supposed to live next to a component in order to easily request all the data needed for your component. Thus, since we need to display some new data on the `ProductSynthesis` component, we update the `ProductSynthesisFragment`.

Reviewer: ```suggestion
GraphQL queries can be split in Fragments and each fragment is supposed to live next to a component in order to easily request all the data needed for your component. Thus, since we need to display some new data on the `ProductSynthesis` component, we update the `ProductSynthesisFragment`.
```

</blockquote>

## Displaying the result on screen

We're almost there! In the file `Synthesis.js` (which generates the product page), you can add some code to display the `rate`. Knowing React.js or at least HTML syntax is necessary to perform this step without impacting other items on the page.

```diff
// Many more things here
const ProductSynthesis = props => {
  const {
    product
    // Many more things here
  } = props;
  return (
    <div className="product__synthesis">
      {/* Many more things here */}
+     <div>Rate: {product.rate} / 5</div>
      {/* Many more things here */}
    </div>
  );
};

export default ProductSynthesis;
```

## Conclusion

Here is a screenshot of the final result on the product page.

![Alt: Product details screenshot with the rate of the product](./assets/product-details-screenshot.png)

> And here is a screenshot of the GraphQL query.

Reviewer: This screenshot does not feel really relevant here.
You could use it at ligne 104 instead of the code examples.

### How to find which files to edit

On your web page, right click on the element you wish to edit, then click on **Inspect Element**. In the inspector, you can look for the `class` that holds the elements you want to edit. Here, the `class` is `product__synthesis`.

![Inspector screenshot demonstrating that when you inspect a product page description, you find the product__sythesis class in the DOM](./assets/inspector-screenshot.png)

Then you can search for that keyword in your `node_modules/front-commerce/src` folder and you will find the right files.

![Synthesis screenshot that shows where to find the files given a `class` value](./assets/synthesis-screenshot.png)

### Overriding the files

In **this** example, the files to be overridden are `ProductSynthesisFragment.gql` and `Synthesis.js` both located at `node_modules/front-commerce/src/web/theme/modules/ProductView/Synthesis/`. Copy and paste them into `mymodule/web/theme/modules/ProductView/Synthesis`.  
See [Extend the theme](/docs/essentials/extend-the-theme.html) for more details.

> 

Reviewer: ```suggestion

Once you have created the new files, **restart your application** to ensure that these are used instead of the core ones.

```
